presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -100,13 +104,77 @@ private PlanNodeCostEstimate getGroupCost(GroupReference groupReference)
 
     private PlanNodeCostEstimate calculateCumulativeCost(PlanNode node)
     {
-        PlanNodeCostEstimate localCosts = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
+        PlanNodeCostEstimate localCost = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
 
         PlanNodeCostEstimate sourcesCost = sourceProvider.getSources(node).stream()
                 .map(this::getCumulativeCost)
                 .reduce(ZERO_COST, PlanNodeCostEstimate::add);
 
-        PlanNodeCostEstimate cumulativeCost = localCosts.add(sourcesCost);
+        PlanNodeCostEstimate cumulativeCost = node.accept(new CostCombiner(), new CostCombiner.Context(localCost, sourcesCost));
+        verify(cumulativeCost != null);
         return cumulativeCost;
     }
+
+    private static class CostCombiner
+            extends PlanVisitor<PlanNodeCostEstimate, CostCombiner.Context>
+    {
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Context context)
+        {
+            return context.localCost.add(context.sourcesCost);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Context context)
+        {
+            PlanNodeCostEstimate cost = context.localCost.add(context.sourcesCost);
+
+            verify(Doubles.compare(context.localCost.getPeakMemoryCost(), context.localCost.getPeakMemoryWhenOutputting()) == 0);
```

Comment:
`Doubles.compare` delegates to `Double.compare`. But you don't need it here. It is needed if deterministic comparison with  `NaN` is needed, or if `-0` to `0` differentiation is needed. I'm not sure if that's something you expect here. I think just simple `context.localCost.getPeakMemoryCost() == context.localCost.getPeakMemoryWhenOutputting()` will work here, since `NaN == NaN => FALSE`, `-0 == 0 => TRUE`

```
public static void main(String[] args)
{
        System.out.println("NaN == NaN: " + (NaN == NaN));
        System.out.println("-0 == 0: " + (-0.0 == 0.0));
}

NaN == NaN: false
-0 == 0: true
```

Reply:
Also please add a meaningful message in case it ever fails.

Reply:
I needed something that will tell "OK" also for "NaN vs NaN".